src/client/signatures/canonical_json/src/ser.rs

```diff
@@ -0,0 +1,486 @@
+use serde::ser::{Serialize};
+use serde_json::ser::{Formatter, CharEscape};
+use std::io::Write;
+use log::{debug, error};
+
+struct JSONFormatter {}
+
+impl Formatter for JSONFormatter {
+    fn write_f64<W: ?Sized>(&mut self, writer: &mut W, value: f64) -> Result<(), std::io::Error> where
+        W: Write,
+    {
+        writer.write(&format_number(value, 10.0_f64.powf(21.0), 10.0_f64.powf(-6.0)).into_bytes())?;
+        Ok(())
+    }
+
+    fn write_char_escape<W: ?Sized>(&mut self, writer: &mut W, char_escape: CharEscape) -> Result<(), std::io::Error> where
+        W: Write, 
+    {
+        match char_escape {
+            CharEscape::Quote => {
+                writer.write(&"\"".to_owned().into_bytes())?; 
+            },
+            CharEscape::ReverseSolidus => {
+                writer.write(&"\\\\".to_owned().into_bytes())?;
+            },
+            CharEscape::LineFeed => {
+                writer.write(&"\n".to_owned().into_bytes())?;
+            },
+            CharEscape::Tab => {
+                writer.write(&"\t".to_owned().into_bytes())?;
+            },
+            CharEscape::CarriageReturn => {
+                writer.write(&"\r".to_owned().into_bytes())?;
+            },
+            _ => { println!("Other char escape"); }
+        }
+        Ok(())
+    }
+
+    fn write_string_fragment<W: ?Sized>(&mut self, writer: &mut W, fragment: &str) -> Result<(), std::io::Error> where
+        W: Write, 
+    {
+        let formatted_string = format!("{}", fragment).escape_default().to_string().replace("\\\"", "\"");
+        writer.write(&format_unicode_in_string(formatted_string).into_bytes())?;
+        Ok(())
+    }
+}
+
+fn format_number(number: f64, lower_bound: f64, upper_bound: f64) -> String {
+
+    debug!("in format_number function");
+
+    if (0.0 < number && number < upper_bound) || (number >= lower_bound) {
+        debug!("converting number to scientific notation");
+        let number_string = format!("{:e}", number);
+
+        let mut prev_char = '\0';
+        let mut formatted_number_string = String::new();
+        for curr_char in number_string.chars() {
+            if prev_char == 'e' && curr_char != '-' {
+                // its a positive exponent
+                formatted_number_string.pop();
+                formatted_number_string.push_str("e+");
+            }
+
+            formatted_number_string.push(curr_char);
+            prev_char = curr_char;
+        }
+
+        return formatted_number_string.replace(r"\.?0+$", "");
+    }
+
+    debug!("returning number without scientifc notation");
+    return format!("{}", number).replace(r"\.?0+$", "");
+}
+
+fn format_unicode_in_string(serialized_string: String) -> String {
```

Comment:
Might as well make the comment be docstring.  Why not just have this take a writer, rather than building and converting strings in the middle?  Folks who want strings can pass in a `mut String::new()`, since it implements `Write`.